### DISCUSSION TITLE
Qwen3.8-27B on one laptop CPU: native C, 8 GB path, and function tools

### DISCUSSION THREAD

**Discussion by shyringo** [QwenLM/Qwen3#1905] [Show and tell]
I maintain [qwen3.8-27b-in-c](https://github.com/shyringo/qwen3.8-27b-in-c), a native C inference engine built specifically for the Qwen3.8-27B hybrid Gated DeltaNet/attention model.

The goal is to make the model genuinely usable on an ordinary laptop CPU:

- one launcher builds the engine, selects weights for available memory, downloads a pinned Unsloth Dynamic V3 GGUF from ModelScope, and starts a conversation;
- the minimum tested IQ1_M path runs in 8 GB RAM (6.01 GiB measured peak RSS);
- the higher-quality Q4_K_M path is selected when at least 20 GiB is available;
- best measured resident IQ1_M generation on an Intel i5-1340P laptop is 0.397 s/token (2.52 token/s);
- v0.4 adds native OpenAI-compatible function tools, parallel calls, tool-result replay, and live SSE with keep-alives, so local applications can use the resident model for tool-driven work without Python or another inference runtime.

```bash
git clone https://github.com/shyringo/qwen3.8-27b-in-c.git
cd qwen3.8-27b-in-c
./qwen38.sh

# Or keep the model behind a local API:
./qwen38.sh --server 8080 --no-thinking
```

The model graph, bounded GGUF reader, recurrent/KV state, sampling, chat runtime, and CPU execution are implemented in this repository. It does not invoke llama.cpp or another engine at runtime. For the same pinned GGUF, the optimized and native baseline paths produce byte-identical full logits; IQ1_M and Q4_K_M are still lossy quantizations of the original checkpoint, so this claim is about runtime fidelity rather than pretending quantization has no quality trade-off.

![Real Qwen3.8-27B output from the native C engine](https://raw.githubusercontent.com/shyringo/qwen3.8-27b-in-c/main/docs/assets/terminal-demo.png)

The README includes the one-command path, exact TTFT/TPOT conditions, memory measurements, architecture/provenance notes, and a benchmark issue form. Reports from other laptop CPUs would be especially useful.
